src/sdi-progress-window.c

```diff
@@ -0,0 +1,203 @@
+    // message is replaced by a shorter one, to avoid the window expanding
+    // and shrinking over and over every time a message changes. But when
+    // a refresh has ended and its progress bar disappears, it is legit to
+    // resize the window to the minimum.
+    gtk_window_set_default_size(GTK_WINDOW(self->main_window), 0, 0);
+  }
+}
+
+static void add_dialog_to_main_window(SdiProgressWindow *self,
+                                      SdiRefreshDialog *dialog) {
+  if (self->main_window == NULL) {
+    self->main_window = GTK_WINDOW(
+        gtk_application_window_new(GTK_APPLICATION(self->application)));
+    gtk_window_set_deletable(self->main_window, FALSE);
+    self->refresh_bar_container =
+        GTK_BOX(gtk_box_new(GTK_ORIENTATION_VERTICAL, 0));
+    gtk_window_set_child(self->main_window,
+                         GTK_WIDGET(self->refresh_bar_container));
+    gtk_window_set_title(GTK_WINDOW(self->main_window), _("Refreshing snaps"));
+    gtk_window_present(GTK_WINDOW(self->main_window));
+    gtk_window_set_default_size(GTK_WINDOW(self->main_window), 0, 0);
+  }
+
+  gtk_box_append(self->refresh_bar_container, GTK_WIDGET(dialog));
+  gtk_widget_set_visible(GTK_WIDGET(dialog), TRUE);
+  // the 'hide-event' is emitted by the dialog when the user clicks on the
+  // 'Hide' button in that progress bar dialog.
+  g_signal_connect_swapped(G_OBJECT(dialog), "hide-event",
+                           (GCallback)remove_dialog_from_main_window, self);
+}
+
+/**
+ * This callback should be connected to the `begin-refresh` signal from a
+ * #sdi_refresh_monitor object. It will create a new window if required, and
+ * insert into it a new #sdi_refresh_dialog with the snap name, snap icon and
+ * progress bar.
+ */
+void sdi_progress_window_begin_refresh(SdiProgressWindow *self,
+                                       gchar *snap_name, gchar *visible_name,
+                                       gchar *icon, gpointer data) {
+  g_autoptr(SdiRefreshDialog) dialog =
+      g_object_ref_sink(sdi_refresh_dialog_new(snap_name, visible_name));
+  if (icon != NULL) {
+    sdi_refresh_dialog_set_icon_image(dialog, icon);
+  }
+  g_hash_table_insert(self->dialogs, (gpointer)g_strdup(snap_name),
+                      g_object_ref(dialog));
+  add_dialog_to_main_window(self, dialog);
+}
+
+/**
+ * This callback should be connected to the `end-refresh` signal from a
+ * #sdi_refresh_monitor object. It will remove the dialog that corresponds
+ * to the specified snap, and close the window if there aren't any more
+ * refresh dialogs.
+ */
+
+void sdi_progress_window_end_refresh(SdiProgressWindow *self, gchar *snap_name,
+                                     gpointer data) {
+  SdiRefreshDialog *dialog =
+      (SdiRefreshDialog *)g_hash_table_lookup(self->dialogs, snap_name);
+  if (dialog != NULL) {
+    remove_dialog_from_main_window(self, dialog);
+  }
+}
+
+/**
+ * This callback should be connected to the `update-progress` signal from a
```

Comment:
Do you mean `refresh-progress`?
```suggestion
 * This callback should be connected to the `refresh-progress` signal from a
```

Reply:
Again, you are right. That was the old name of the signal. Clearly I forgot to change the references in the comments when I renamed it.